### PR TITLE
Prefer CI info over git if present

### LINKED ISSUE #189: Jenkins with Github Pull Request Builder fails to accurately report
Jenkins when used with Github Pull Request Builder uses a special branch. It's ref is created by Github. 

It looks like this `pull/ID/head` which is the head of your Pull Request or `pull/ID/merge` which is your PR merged into your intended branch. The result is a sha that differs than what you find in your pull request. 

Because it still has a full working git repo in the checkout though, and ruby-test-reporter gives that precedence over info in environment variables coverage is reported with the wrong sha. 

>https://help.github.com/articles/checking-out-pull-requests-locally/
>https://stackoverflow.com/questions/27567846/how-can-i-check-out-a-github-pull-request

### PR DESCRIPTION
**Pull request by andrewhood125**
Resolves #189

### CODE REVIEW

File: spec/code_climate/test_reporter/git_spec.rb [hunk 1]
```diff
@@ -79,18 +79,18 @@ def self.root
...
         expect(Git).to receive(:git).with("log -1 --pretty=format:'%H'").and_return("1234")
 
-        expect(Git.head_from_git_or_ci).to eq '1234'
+        expect(Git.head_from_ci_or_git).to eq '1234'
       end
 
-      it 'returns the head sha from ci if git is not available' do
-        expect(Git).to receive(:git).with("log -1 --pretty=format:'%H'").and_return("")
-        expect(Ci).to receive(:service_data).and_return({commit_sha: "4567"})
+      it 'returns the head sha from git if ci is not available' do
```
Comment (maxjacobson) [inline spec/code_climate/test_reporter/git_spec.rb]: it looks like we're missing a spec `it 'returns the head from from CI when both git and ci are available'`

File: lib/code_climate/test_reporter/git.rb [hunk 2]
```diff
@@ -44,7 +44,8 @@ def head_from_git
         end
 
         def head_from_ci
-          Ci.service_data[:commit_sha]
+          sha = Ci.service_data[:commit_sha]
+          return nil if sha.empty?
```
Comment (maxjacobson) [inline lib/code_climate/test_reporter/git.rb]: as-written, this will always return `nil`

**Comment by andrewhood125** [comment]
@maxjacobson thanks for pointing me to that! I'll take a look!

File: lib/code_climate/test_reporter/git.rb (hunk 2 above)
Comment (andrewhood125) [inline lib/code_climate/test_reporter/git.rb]: Right you are. So weird.. though I ran this from `pry` and it returned the sha. I'll be explicit about it. 👍

**Comment by andrewhood125** [comment]
@maxjacobson updated! 👍

**Comment by andrewhood125** [comment]
Hmm.. looking at fails.

**Comment by andrewhood125** [comment]
No safe operator. 😢

**Comment by andrewhood125** [comment]
Looks like for some reason the ordering of the has was switched. I wonder if a less strict compare would be better. Thoughts @maxjacobson ?

**Comment by maxjacobson** [comment]
I think I know what's happening. The expectation that's not met is here:

```
       -:git => {:head=>"7a36651c654c73e7e9a6dfc9f9fa78c5fe37241e", :committed_at=>1474318896, :branch=>"master"},
       +:git => {:head=>"73a72f2142b1dbba41cefa0181b996d856f7ee4b", :committed_at=>1474318896, :branch=>"master"},
```

The `head` is expected to be the latest sha in this compressed repository https://github.com/codeclimate/ruby-test-reporter/blob/master/spec/fixtures/fake_project.tar.gz , but instead, when the test runs on Circle CI, it prefers the value from the environment. When you run the test locally, no CI-specific environment variables are set, so it uses the one from the repository.

Does that make sense? I think you can probably avoid this failure by stubbing something.

**Comment by andrewhood125** [comment]
Ah I see it now. I glazed over those shas think they were the same 👍 Thanks for highlighting that.

**Comment by andrewhood125** [comment]
@maxjacobson updated.

**Comment by andrewhood125** [comment]
@maxjacobson look good to you?

**Comment by maxjacobson** [comment]
@andrewhood125 hey, thanks for the ping and for working on this. I was about to merge this, but then noticed that my colleague was making [a similar change in our new test reporter](https://github.com/codeclimate/test-reporter/pull/180). After talking it through, I believe this change isn't as thorough as it needs to be. Before merging, we would need to also update this to prefer the CI information for `branch` and `committed_at`.

I do appreciate your work, and if you'd like to make that change I'll be happy to merge it. But if you're interested, what do you think about trying out our new test reporter (it's in beta and has [some additional cool features](https://codeclimate.com/changelog/591c99570d51f6481f001291))? As of today's release, it should already work for your use-case.

**Comment by andrewhood125** [comment]
👌 got the new test reporter working.

**Pull request closed without merging**